File: services/cms-svc/src/main/java/com/supportme/pay/cmssvc/career/JobApplication.java
package com.supportme.pay.cmssvc.career;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import org.hibernate.annotations.CreationTimestamp;

import java.time.OffsetDateTime;

/**
 * Zgłoszenie rekrutacyjne. jobPositionId = null oznacza aplikację
 * spontaniczną — ta MIMO TO ma właściciela (organizationId ustawiany
 * niezależnie od stanowiska).
 */
@Entity
@Table(name = "job_applications")
public class JobApplication {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "job_position_id")
    private Long jobPositionId;

    @Column(name = "organization_id")
    private Long organizationId;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    private String email;

    @Column
    private String phone;

    @Column(columnDefinition = "text")
    private String message;

    @Column(name = "cv_path")
    private String cvPath;

    @Column(name = "cv_original_name")
    private String cvOriginalName;

    @Column(name = "is_read", nullable = false)
    private boolean isRead = false;

    @Column(nullable = false, length = 20)
    private String status = "pending";

    @Column(name = "future_recruitment_consent", nullable = false)
    private boolean futureRecruitmentConsent = false;

    @Column(name = "future_recruitment_consent_at")
    private OffsetDateTime futureRecruitmentConsentAt;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    /** Wymagany przez Hibernate. */
    protected JobApplication() {
    }

    public JobApplication(Long jobPositionId, Long organizationId, String name, String email, String phone,
                           String message, String cvPath, String cvOriginalName, boolean futureRecruitmentConsent) {
        this.jobPositionId = jobPositionId;
        this.organizationId = organizationId;
        this.name = name;
        this.email = email;
        this.phone = phone;
        this.message = message;
        this.cvPath = cvPath;
        this.cvOriginalName = cvOriginalName;
        this.futureRecruitmentConsent = futureRecruitmentConsent;
        this.futureRecruitmentConsentAt = futureRecruitmentConsent ? OffsetDateTime.now() : null;
    }

    public Long getId() {
        return id;
    }

    public Long getJobPositionId() {
        return jobPositionId;
    }

    public Long getOrganizationId() {
        return organizationId;
    }

    public String getName() {
        return name;
    }

    public String getEmail() {
        return email;
    }

    public String getPhone() {
        return phone;
    }

    public String getMessage() {
        return message;
    }

    public String getCvPath() {
        return cvPath;
    }

    public String getCvOriginalName() {
        return cvOriginalName;
    }

    public boolean isRead() {
        return isRead;
    }

    public void setRead(boolean read) {
        isRead = read;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public boolean isFutureRecruitmentConsent() {
        return futureRecruitmentConsent;
    }

    public OffsetDateTime getFutureRecruitmentConsentAt() {
        return futureRecruitmentConsentAt;
    }

    public OffsetDateTime getCreatedAt() {
        return createdAt;
    }
}

File: pay-backend/storefront-domain/src/main/kotlin/com/supportme/pay/storefront/domain/entity/JobApplication.kt
package com.supportme.pay.storefront.domain.entity

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table
import org.hibernate.annotations.CreationTimestamp
import java.time.Instant
import java.time.ZoneOffset

/**
 * Zgłoszenie rekrutacyjne. `position == null` = aplikacja spontaniczna.
 * `futureRecruitmentConsent*` — zgoda GDPR na przyszłe procesy rekrutacyjne,
 * ważna [FUTURE_CONSENT_MONTHS] miesięcy od udzielenia (odpowiednik
 * `JobApplication::futureConsentExpiresAt()`/`futureConsentActive()`).
 */
@Entity
@Table(name = "job_applications")
class JobApplication(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "job_position_id")
    var position: JobPosition? = null,

    @Column(nullable = false)
    var name: String,

    @Column(nullable = false)
    var email: String,

    @Column
    var phone: String? = null,

    @Column(columnDefinition = "text")
    var message: String? = null,

    @Column(name = "cv_path")
    var cvPath: String? = null,

    @Column(name = "cv_original_name")
    var cvOriginalName: String? = null,

    @Column(name = "is_read", nullable = false)
    var isRead: Boolean = false,

    @Column(nullable = false)
    var status: JobApplicationStatus = JobApplicationStatus.PENDING,

    @Column(name = "future_recruitment_consent", nullable = false)
    var futureRecruitmentConsent: Boolean = false,

    @Column(name = "future_recruitment_consent_at")
    var futureRecruitmentConsentAt: Instant? = null,

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    val createdAt: Instant? = null,
) {
    /** Arytmetyka kalendarzowa (miesiące, nie dni stałej długości) — jak `Carbon::addMonths()`. */
    fun futureConsentExpiresAt(): Instant? {
        if (!futureRecruitmentConsent || futureRecruitmentConsentAt == null) return null
        return futureRecruitmentConsentAt!!.atZone(ZoneOffset.UTC).plusMonths(FUTURE_CONSENT_MONTHS.toLong()).toInstant()
    }

    fun futureConsentActive(): Boolean {
        val expiry = futureConsentExpiresAt() ?: return false
        return expiry.isAfter(Instant.now())
    }

    companion object {
        const val FUTURE_CONSENT_MONTHS = 24
    }
}
